ambry-account/src/main/java/com/github/ambry/account/HelixAccountService.java

```diff
@@ -0,0 +1,534 @@
+      throw new IllegalArgumentException(
+          "storeConfig=" + storeConfig + " and storeFactory=" + storeFactory + " cannot be null");
+    }
+    long startTimeMs = System.currentTimeMillis();
+    logger.info("Starting HelixAccountService with HelixPropertyStore host={} rootPath={}",
+        storeConfig.zkClientConnectString, storeConfig.rootPath);
+    helixStore = storeFactory.getHelixPropertyStore(storeConfig, null);
+    readFullAccountAndUpdateCache(FULL_ACCOUNT_METADATA_PATH);
+    isOpen = true;
+    logger.info("HelixAccountService started. {} accounts have been loaded, took time={}ms",
+        cache.getAccountTwoKeyMap().getAccounts().size(), System.currentTimeMillis() - startTimeMs);
+  }
+
+  @Override
+  public Account getAccountByName(String accountName) {
+    if (!isOpen) {
+      throw new IllegalStateException("AccountService is closed.");
+    }
+    if (accountName == null) {
+      throw new IllegalArgumentException("accountName cannot be null.");
+    }
+    return cache.getAccountByName(accountName);
+  }
+
+  @Override
+  public Account getAccountById(short id) {
+    if (!isOpen) {
+      throw new IllegalStateException("AccountService is closed.");
+    }
+    return cache.getAccountById(id);
+  }
+
+  @Override
+  public Collection<Account> getAllAccounts() {
+    if (!isOpen) {
+      throw new IllegalStateException("AccountService is closed.");
+    }
+    return cache.getAllAccounts();
+  }
+
+  /**
+   * {@inheritDoc}
+   * <p>
+   *   This call is blocking until it completes the operation of updating account metadata to {@link HelixPropertyStore}.
+   * </p>
+   * <p>
+   *   There is a slight chance that an {@link Account} could be updated successfully, even if its value was
+   *   set based on an outdated {@link Account}. This can be fixed after {@code generationId} is introduced
+   *   to {@link Account}.
+   * </p>
+   */
+  @Override
+  public boolean updateAccounts(Collection<Account> accounts) {
+    if (!isOpen) {
+      throw new IllegalStateException("AccountService is closed.");
+    }
+    if (accounts == null) {
+      throw new IllegalArgumentException("accounts cannot be null");
+    }
+    long startTimeMs = System.currentTimeMillis();
+    logger.trace("Start updating to HelixPropertyStore with accounts={}", accounts);
+    if (hasDuplicateAccountIdOrName(accounts)) {
+      return false;
+    }
+    boolean hasSucceeded = false;
+    // make a pre check for conflict between the accounts to update and the accounts in the local cache. Will fail this
+    // update operation if any conflict exists. There is a slight chance that the account to update conflicts with
+    // the accounts in the local cache, but does not conflict with those in the helixPropertyStore. This will happen
+    // if some accounts are updated but the local cache is not refreshed.
+    if (!hasConflictWithCache(accounts)) {
+      hasSucceeded = helixStore.update(FULL_ACCOUNT_METADATA_PATH, currentData -> {
```

Comment:
Guess we could not. This is the map fetched from remote. Will come to you and explain.